### PR TITLE
Add CORS header to avoid browser warnings

### PR DESCRIPTION
**Pull request by KarlStraussberger**
CORS warnings happen e.g. when using gerbera web ui with dns name

### CODE REVIEW

**Comment by Tieske** [comment]
this will prevent browsers from sending any credentials iirc, could that cause problems?

**Comment by KarlStraussberger** [comment]
> this will prevent browsers from sending any credentials iirc, could that cause problems?

If CORS kicks in the browser blocks communication to the server, including log in.

**Comment by KarlStraussberger** [comment]
The browers send credentials as soon as there is a request in a script. CORS just blocks scripts from being executed. If we have a wildcard we allow scripts beeing loaded from any source.
Of course, a new library function to overwrite the value would be best, but this brings loads of version checks in applications whether the function is available or not.

**Comment by Tieske** [comment]
From this link: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors/CORSNotSupportingCredentials

What I read is that a browser cannot access any resource on the (upnp) server while providing credentials. So if there are security scheme's in effect (there are some UPnP docs on security, but again dunno the details), then with this change, those might break right?

Not saying this won't work, just wondering if there is a risk of breaking anything.

**Comment by KarlStraussberger** [comment]
Wasn't aware of that restriction. As a consequence we have to add a method to set that value, which still leaves us with the problem that older binaries running with newer libraries will have to run without CORS being set at all.

File: upnp/src/genlib/net/http/webserver.c [hunk 1]
```diff
@@ -1367,6 +1367,15 @@ static int process_request(
 		/*          goto error_handler; */
 		/*      } */
 	}
+	{
+		UpnpListHead *eHeader =
+			UpnpFileInfo_get_ExtraHeadersList(finfo);
+		UpnpExtraHeaders *h = UpnpExtraHeaders_new();
+		UpnpExtraHeaders_set_resp(h, "Access-Control-Allow-Origin: *");
```
Comment (randomstuff) [inline upnp/src/genlib/net/http/webserver.c]: Is this allowing *any server anywhere on the web* to access the UPnP resources? I that's the case  I don't believe that's desirable.

(I am not sure what is supposed to happen when the other `Access-Control-Allow-*` are absent).

Comment (KarlStraussberger) [reply upnp/src/genlib/net/http/webserver.c]: It's mainly controlling the access of scripts from other sources, a.k.a side-loading

**Comment by KarlStraussberger** [comment]
That's completely fine. Tried to upload the files instead of creating a local branch - not a good idea.

**Comment by mrjimenez** [comment]
@KarlStraussberger ,

When you feel your are done with your work, please tell me so that I can do a new release.

No hurries on my side.

Best regards,
Marcelo.

**Comment by KarlStraussberger** [comment]
Will do the local branch tomorrow to make sure all is set and done.

**Comment by KarlStraussberger** [comment]
@mrjimenez No more findings so you can go on as you like.

File: upnp/src/genlib/net/http/webserver.c (hunk 1 above)
Comment (randomstuff) [reply upnp/src/genlib/net/http/webserver.c]: Won't this make it easier for a remote web server to scan your LAN (or local machine):

- for fingerprinting your LAN/device (eg. when you are using a VPN),
- for detecting the IP address, port and version of UPnP software in order to try to trigger an attack?

**Comment by randomstuff** [comment]
Moved this comment here for more visibility.

Won't this make it easier for a remote web server to scan your LAN (or local machine):

    for fingerprinting your LAN/device (eg. when you are using a VPN),
    for detecting the IP address, port and version of UPnP software in order to try to trigger an attack?

**Comment by KarlStraussberger** [comment]
If you add a "*" yes. That's why I revoked the change and allow to set it to values by the upnp server (e.g. gerbera).

We will allow calls from 'self' virtualHostname and externalHostname only to make sure scripts are loaded when running behind a proxy or in a container.

**Comment by Vollstrecker** [comment]
> When you feel your are done with your work, please tell me so that I can do a new release.

Maybe wait with the release until I'm sure (https://github.com/amule-project/amule/issues/374) is not caused by the lib.

**Pull request closed without merging**